### PR TITLE
[DI] Introduce "container.service_locator" tag, replaces ServiceLocatorArgument

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no (master only)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

I first started working on adding this new "container.service_locator" tag, so here it is.
It allows defining and dumping service-locator services properly, where it wasn't possible previously (you had to create a DI extension to do so.)

Then I realized that this allowed us to entirely drop `ServiceLocatorArgument` and replace it with the more flexible `ServiceClosureArgument`.

This makes things simpler overall, see diff stat.

### CODE REVIEW

**Review by chalasr — APPROVED** [review]
Makes good sense to me 👍

**Review by GuilhemN — APPROVED** [review]
:+1:

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml [hunk 1]
```diff
@@ -49,8 +49,13 @@
 
         <service id="session_listener" class="Symfony\Component\HttpKernel\EventListener\SessionListener">
             <tag name="kernel.event_subscriber" />
-            <argument type="service-locator">
-                <argument key="session" type="service" id="session" on-invalid="null" />
+            <argument type="service">
+                <service class="Symfony\Component\DependencyInjection\ServiceLocator">
+                    <tag name="service_locator" />
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml]: Almost all other tags are of the "X.Y" form. Not sure if we need to call it `container.service_locator`.

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml]: tag renamed to container.service_locator

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by sstok** [comment]
If you always use `Definition` to define a ServiceLocator would it make sense to add `ServiceLocatorDefinition` of sorts? So you don't have to remember adding the tag every time.

**Comment by nicolas-grekas** [comment]
@sstok adding `ServiceLocatorDefinition` would create a new type, not required to me. A helper method would the job equally well. But right now, I don't feel the need for one personally.

**Pull request merged**